### PR TITLE
Fix skipping in parquet int64 millis decoder

### LINKED ISSUE #5443: GENERIC_INTERNAL_ERROR (65536) java.lang.UnsupportedOperationException: undefined
We have presto 341 Running , Getting below error when trying to access a Glue table using hive connector. The table is in parquet format

This does not exist on v340 but the same behavior is there in version 343.

```
java.lang.UnsupportedOperationException: undefined
	at io.prestosql.parquet.dictionary.Dictionary.decodeToBinary(Dictionary.java:33)
	at io.prestosql.parquet.dictionary.DictionaryReader.readBytes(DictionaryReader.java:53)
	at io.prestosql.parquet.reader.Int64TimestampMillisColumnReader.skipValue(Int64TimestampMillisColumnReader.java:54)
	at io.prestosql.parquet.reader.PrimitiveColumnReader.processValues(PrimitiveColumnReader.java:203)
	at io.prestosql.parquet.reader.PrimitiveColumnReader.skipValues(PrimitiveColumnReader.java:191)
	at io.prestosql.parquet.reader.PrimitiveColumnReader.seek(PrimitiveColumnReader.java:232)
	at io.prestosql.parquet.reader.PrimitiveColumnReader.readPrimitive(PrimitiveColumnReader.java:162)
	at io.prestosql.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:262)
	at io.prestosql.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:314)
	at io.prestosql.parquet.reader.ParquetReader.readBlock(ParquetReader.java:297)
	at io.prestosql.plugin.hive.parquet.ParquetPageSource$ParquetBlockLoader.load(ParquetPageSource.java:164)
	at io.prestosql.spi.block.LazyBlock$LazyData.load(LazyBlock.java:381)
	at io.prestosql.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:360)
	at io.prestosql.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:276)
	at io.prestosql.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.setupDictionaryBlockProjection(DictionaryAwarePageProjection.java:211)
	at io.prestosql.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.lambda$getResult$0(DictionaryAwarePageProjection.java:197)
	at io.prestosql.spi.block.LazyBlock$LazyData.load(LazyBlock.java:381)
	at io.prestosql.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:360)
	at io.prestosql.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:276)
	at io.prestosql.operator.project.PageProcessor$ProjectSelectedPositions.processBatch(PageProcessor.java:342)
	at io.prestosql.operator.project.PageProcessor$ProjectSelectedPositions.process(PageProcessor.java:205)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1076)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_341____20201007_074354_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by martint**
Fixes https://github.com/prestosql/presto/issues/5443

### CODE REVIEW

**Review by sopel39 — APPROVED** [review]
Would test be possible?

**Comment by martint** [comment]
> Would test be possible?

I'm looking into it

**Comment by martint** [comment]
@sopel39, added test

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import static org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory.javaLongObjectInspector;
+import static org.apache.parquet.column.ParquetProperties.WriterVersion.PARQUET_1_0;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.COMPRESSION;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.ENABLE_DICTIONARY;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.WRITER_VERSION;
+import static org.apache.parquet.schema.MessageTypeParser.parseMessageType;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+public class TestReader
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:52]: its not clear what belongs to TestReader and what to TestParquetReader (same package)

Comment (martint) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:52]: Couldn't come up with a better name, and TestParquetReader is taken (it's the full blown test with all the combinations of readers and writers). This is a targeted test for one type. I could rename it to TestTimestampReader for now -- I'll have to revamp the TestParquetReader for the new parquet reader I'm working on, anyway, as the current one is too rigid to be able to implement the tests I need.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/ParquetTester.java [hunk 2]
```diff
@@ -569,7 +569,7 @@ private static FileFormat getFileFormat()
         return OPTIMIZED ? FileFormat.PRESTO_PARQUET : FileFormat.HIVE_PARQUET;
     }
 
-    private static void writeParquetColumn(
+    public static void writeParquetColumn(
```
Comment (sopel39) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/ParquetTester.java:572]: make it package-private?

**Review by sopel39 — APPROVED** [review]

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        for (long value : epochMillisValues) {
+            timestamps.add(SqlTimestamp.fromMillis(3, value));
+        }
+
+        List<ObjectInspector> objectInspectors = singletonList(javaLongObjectInspector);
+        List<String> columnNames = ImmutableList.of("test");
+
+        ConnectorSession session = getHiveSession(new HiveConfig()
+                .setHiveStorageFormat(HiveStorageFormat.PARQUET)
+                .setUseParquetColumnNames(true));
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:70]: None of these is needed

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        try (ParquetTester.TempFile tempFile = new ParquetTester.TempFile("test", "parquet")) {
+            JobConf jobConf = new JobConf();
+            jobConf.setEnum(COMPRESSION, CompressionCodecName.SNAPPY);
+            jobConf.setBoolean(ENABLE_DICTIONARY, true);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:75]: seems unrelated?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                    new Iterator<?>[] {epochMillisValues.iterator()},
+                    Optional.of(parquetSchema),
+                    false);
+
+            Iterator<SqlTimestamp> expectedValues = timestamps.build().iterator();
+            try (ConnectorPageSource pageSource = FileFormat.PRESTO_PARQUET.createFileFormatReader(session, HDFS_ENVIRONMENT, tempFile.getFile(), columnNames, ImmutableList.of(TIMESTAMP_MILLIS))) {
+                // skip a page to exercise the decoder's skip() logic
+                Page firstPage = pageSource.getNextPage();
+                for (int i = 0; i < firstPage.getPositionCount(); i++) {
+                    expectedValues.next();
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:93]: assert firstPage.getPositionCount() > 0

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                }
+
+                while (!pageSource.isFinished()) {
+                    Page page = pageSource.getNextPage();
+                    if (page == null) {
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:98]: We should document what null return value means. "This method is allowed to return null" is not sufficient.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                    Block block = page.getBlock(0);
+
+                    for (int i = 0; i < block.getPositionCount(); i++) {
+                        assertEquals(TIMESTAMP_MILLIS.getObjectValue(session, block, i), expectedValues.next());
+                    }
+                }
+            }
+
+            assertFalse(expectedValues.hasNext());
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:109]: add a message, othewise exception coming from here will be hard to understand

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import static org.testng.Assert.assertFalse;
+
+public class TestReader
+{
+    @Test
+    public void testTimestampBackedByInt64()
+            throws Exception
+    {
+        MessageType parquetSchema = parseMessageType("message hive_timestamp { optional int64 test (TIMESTAMP_MILLIS); }");
+        ContiguousSet<Long> epochMillisValues = ContiguousSet.create(Range.closedOpen((long) -1_000, (long) 1_000), DiscreteDomain.longs());
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:59]: Is negative value any special?
is the range length (2000) special? Can it be eg 20?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/ParquetTester.java (hunk 2 above)
Comment (martint) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/ParquetTester.java:572]: It's now part of the public API of this class, so package private doesn't seem appropriate.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java (hunk 8 above)
Comment (martint) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:59]: No, nothing special about it, other than that it has to be large enough to produce multiple pages. I added some assertions to ensure the assumptions of the test hold.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java (hunk 6 above)
Comment (martint) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestReader.java:98]: Agreed, but unrelated to this PR

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestTimestampReader.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                    false);
+
+            Iterator<SqlTimestamp> expectedValues = timestamps.build().iterator();
+            try (ConnectorPageSource pageSource = FileFormat.PRESTO_PARQUET.createFileFormatReader(session, HDFS_ENVIRONMENT, tempFile.getFile(), columnNames, ImmutableList.of(TIMESTAMP_MILLIS))) {
+                // skip a page to exercise the decoder's skip() logic
+                Page firstPage = pageSource.getNextPage();
+
+                assertThat(firstPage.getPositionCount())
+                        .withFailMessage("Expected first page to have at least 1 row")
+                        .isGreaterThan(0);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestTimestampReader.java:87]: the shorter, intermediate version that you had
```
assertTrue(firstPage.getPositionCount() > 0, "Expected first page to have at least 1 row");
```

produces equally informative exception message (assuming page.positionCount is non-negative)

**Review by findepi — APPROVED** [review]

File: presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestTimestampReader.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static java.util.Collections.singletonList;
+import static org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory.getStandardStructObjectInspector;
+import static org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory.javaLongObjectInspector;
+import static org.apache.parquet.column.ParquetProperties.WriterVersion.PARQUET_1_0;
+import static org.apache.parquet.hadoop.ParquetOutputFormat.WRITER_VERSION;
+import static org.apache.parquet.schema.MessageTypeParser.parseMessageType;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestTimestampReader
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestTimestampReader.java:48]: TestTimestampColumnReader ?

Comment (findepi) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/parquet/TestTimestampReader.java:48]: or, if  you reserve that name for unit tests of that class, than maybe `TestTimestamp`?

**Pull request merged**